pkg/kube/cluster-init.sh

```diff
@@ -255,30 +250,61 @@ longhorn_post_install_config() {
 }
 
 check_start_k3s() {
+  # the cluster change code is in another task loop, so if the cluster wait is nogoing
+  # don't go to start k3s in this time. wait also
+  if [ -f "$CLUSTER_WAIT_FILE" ]; then
+        logmsg "Cluster wait ongoing, wait for it before starting k3s"
+        while [ -f "$CLUSTER_WAIT_FILE" ]; do
+                sleep 5
+        done
+  fi
+
   pgrep -f "k3s server" > /dev/null 2>&1
   if [ $? -eq 1 ]; then
-      if [ $RESTART_COUNT -lt $MAX_K3S_RESTARTS ]; then
-          ## Must be after reboot, or from k3s restart
-          RESTART_COUNT=$((RESTART_COUNT+1))
-          save_crash_log "k3s.log"
-          ln -s /var/lib/k3s/bin/* /usr/bin
-          logmsg "Starting k3s server, restart count: $RESTART_COUNT"
-          # for now, always copy to get the latest
-          nohup /usr/bin/k3s server --config /etc/rancher/k3s/config.yaml &
-          k3s_pid=$!
-          # Give the embedded etcd in k3s priority over io as its fsync latencies are critical
-          ionice -c2 -n0 -p $k3s_pid
-          # Default location where clients will look for config
-          # There is a very small window where this file is not available
-          # while k3s is starting up
-          while [ ! -f /etc/rancher/k3s/k3s.yaml ]; do
-            sleep 5
-          done
-          ln -s /etc/rancher/k3s/k3s.yaml ~/.kube/config
-          mkdir -p /run/.kube/k3s
-          cp /etc/rancher/k3s/k3s.yaml /run/.kube/k3s/k3s.yaml
-          return 1
-      fi
+        # do exponential backoff for k3s restart, but not more than MAX_WAIT_TIME
+        RESTART_COUNT=$((RESTART_COUNT+1))
+        logmsg "k3s server not running, restart wait time $current_wait_time, restart count: $RESTART_COUNT"
+        sleep $current_wait_time
+        current_wait_time=$((current_wait_time * 2))
+        if [ $current_wait_time -gt $MAX_WAIT_TIME ]; then
+                current_wait_time=$MAX_WAIT_TIME
+        fi
+
+        ## Must be after reboot, or from k3s restart
+        save_crash_log
+        ln -s /var/lib/k3s/bin/* /usr/bin
+        if [ ! -d /var/lib/cni/bin ] || [ ! -d /opt/cni/bin ]; then
+                copy_cni_plugin_files
+        fi
+        # for now, always copy to get the latest
+
+        # start the k3s server now
+        nohup /usr/bin/k3s server --config "$k3s_config_file" &
+
+        # remember the k3s start time
+        k3s_last_start_time=$(date +%s)
```

Comment:
Did you intend to use this variable?

Reply:
removed, no need for this in the latest change.